_drafts/tracking-code-size.md

```diff
@@ -0,0 +1,571 @@
+listed in the [GCC flags
+post]({% post_url 2019-08-20-code-size-optimization-gcc-flags %}), this is a
+good place to start.
+
+## Keeping the Code Size to a Minimum
+
+Up until now, the code sizes we've analyzed had all already been committed to
+master. If our firmware is out of space today, we'd have to retro-actively
+refactor or revert these changes which are difficult and time-consuming.
+
+However, if you haven't run out of space yet, or you are looking to keep the
+future code space bloat to a minimum, you need to ensure that every new commit
+and its code size impact is validated.
+
+The best way I've found to keep code size in check is to calculate how much each
+pull-request contributes (or removes!) from the total. Now that we most of the
+data we need, this is possible!
+
+### Code Size Deltas on Pull Requests
+
+To get deltas on pull requests, we'll need CI to run a script whenever one is
+published or updated. This script will take a local binary, compute its code
+size, and compare it against a given revision or the
+`git-merge-base`[^git_merge_base] of the "master" branch. `git-merge-base` takes
+both revisions given and finds the most recent common ancestor between the two.
+In the case of the Zephyr Project, this is typically finding the parent commit
+between the base of a pull-request and the `origin/master` branch.
+
+In our
+[example code](https://github.com/memfault/interrupt/tree/master/example/tracking-code-size/codesizes.py),
+we have a command, `diff` which will do exactly this.
+
+```python
+@cli.command()
+@click.argument("binary")
+@click.option("--build", default="default")
+@click.option("--revision")
+def diff(binary, build, revision):
+    """
+    Calculate the size of the BINARY given, and compare it
+    to either the REVISION provided or the "merge-base" of the
+    current branch and master for the given BUILD.
+    """
+    text, data, bss = _calculate_codesize(binary)
+    if not revision:
+        click.echo("Using `git merge-base` against master branch")
+        # git merge-base HEAD master
+        revision = _get_merge_base("HEAD", "master")
+
+    codesize = _fetch_codesize(build, revision)
+    if not codesize:
+        click.echo("No code size for build: {}, revision: {}".format(
+            build, revision))
+        return
+
+    text_delta = text - codesize.text
+    data_delta = data - codesize.data
+    bss_delta = bss - codesize.bss
+
+    click.echo("Comparing current binary against: {}".format(revision))
+    click.echo(dict(text=text_delta, data=data_delta, bss=bss_delta))
+```
+
+Let's try it out! I went digging into the Zephyr Github Pull Requests page for a
+large changeset, and found a good candidate:
+[Timeout API rework #23382](https://github.com/zephyrproject-rtos/zephyr/pull/23382).
+
+We'll check out the branch and build the firmware first.
+
+```
+# Check out the branch for #23382
+$ cd zephyrproject/zephyr
+$ git remote add pr git@github.com:andyross/zephyr.git
```

Comment:
Maybe: Probably not super important but could copy and/or fork this on memfault/ just to keep it around